### PR TITLE
Remove @Inject from rest handlers

### LINKED ISSUE #15685: Remove guice from Rest*Action
If we're going to remove `@Inject` from the node level stuff we have to start somewhere. I will start with `Rest*Action` classes because it looks reasonably easy.

### PR DESCRIPTION
**Pull request by nik9000**
Do this by adding a single @Inject to NetworkModule. This gives us
a string to pull on: move parameters from injected to constructor.

This is an incremental step to cleaning up our reliance on guice and should
be minimally invasive.

Closes #15685

### CODE REVIEW

**Comment by nik9000** [comment]
I'm still running tests now. On a slow machine today....

**Comment by rjernst** [comment]
We just got rid of RestModule! Why can't the rest actions be created in NetworkModule?

**Comment by nik9000** [comment]
> Why can't the rest actions be created in NetworkModule?

I guess it could. I pulled them out initially because I wanted to work on rest stuff in isolation from the transport stuff. I can smoosh it back together but I kind of like rest stuff registered in a RestModule and transport stuff NetworkModule. NetworkModule just looks so much more cohesive now.

**Comment by rjernst** [comment]
> I kind of like rest stuff registered in a RestModule and transport stuff NetworkModule

The entire point of the refactoring that removed RestModule was Rest is really network. Transport and rest requests aren't really any different, they all have to do with communicating with ES over the network (and while the transport actions are not yet registered in NetworkModule, I mentioned in the PR that it was my intention to do that in a follow up).

**Comment by s1monw** [comment]
> The entire point of the refactoring that removed RestModule was Rest is really network. Transport and rest requests aren't really any different, they all have to do with communicating with ES over the network (and while the transport actions are not yet registered in NetworkModule, I mentioned in the PR that it was my intention to do that in a follow up).

while I agree I think we can do this in a followup? This is already awesome and worth a push as it is? :)

**Comment by s1monw** [comment]
oh nevermind, now I see it's adding RestModule back, yeah I think we should never ever again add modules back, can you just fold it into networkModule?

File: core/src/main/java/org/elasticsearch/rest/RestModule.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        registerSearchTemplateActions();
+        registerSettingsActions();
+        registerSnapshotActions();
+        registerWarmerActions();
+    }
+
+    public <T extends BaseRestHandler> void add(Class<T> type, Function<RestGlobalContext, T> builder) {
+        requireNonNull(type, "Must define the handler type being registered");
+        requireNonNull(builder, "Must define the builder to register");
+        Object old = actions.put(type, builder);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestModule.java]: nit: use `putIfAbsent` then you don't modify even if it's there?

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/node/hotthreads/RestNodesHotThreadsAction.java [hunk 2]
```diff
@@ -40,19 +38,17 @@
 /**
  */
 public class RestNodesHotThreadsAction extends BaseRestHandler {
+    public RestNodesHotThreadsAction(RestGlobalContext context) {
+        super(context);
+        context.getController().registerHandler(RestRequest.Method.GET, "/_cluster/nodes/hotthreads", this);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/node/hotthreads/RestNodesHotThreadsAction.java]: can we maybe have a final registerHandler method on `BaseRestHandler`?  eventually we should do this also in the module but that's a lot of work

**Comment by nik9000** [comment]
> oh nevermind, now I see it's adding RestModule back, yeah I think we should never ever again add modules back, can you just fold it into networkModule?

I'll do it but I don't get it. NetworkModule without the rest stuff it just sets up the transport implementations and leaves the ticklish bits to something else. But its a separate thing. I'll put it back.

File: core/src/main/java/org/elasticsearch/rest/RestModule.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/RestModule.java]: Done.

File: core/src/main/java/org/elasticsearch/rest/action/admin/cluster/node/hotthreads/RestNodesHotThreadsAction.java (hunk 2 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/rest/action/admin/cluster/node/hotthreads/RestNodesHotThreadsAction.java]: I just pushed a patch along those lines. Its not what you asked for but its something....

**Comment by nik9000** [comment]
I believe I've addressed all the comments, @s1monw.

**Comment by nik9000** [comment]
Rebased. Sorry for the trouble.

**Comment by nik9000** [comment]
I'm going to rebase again. Any interest in another review?

**Comment by nik9000** [comment]
Rebased!

**Comment by nik9000** [comment]
This is super out of date now.

**Pull request closed without merging**